fs_test.go

```diff
@@ -44,6 +45,186 @@ func TestMain(m *testing.M) {
+		fixtures := []struct {
+			desc   string
+			offset int64
+			whence int
+		}{
+			{"seek before beginning", -1, io.SeekCurrent},
+			{"seek after end", int64(len(content)) + 1, io.SeekStart},
+			{"seek exactly to end", 0, io.SeekEnd},
+		}
+
+		for _, f := range fixtures {
+			f := f
+			t.Run(f.desc, func(t *testing.T) {
+				testFs := s3fs.New(s3cl, *bucket)
+				data, err := testFs.Open(testFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				_, err = data.(io.Seeker).Seek(f.offset, f.whence)
+				if err == nil {
+					t.Fatal(err)
+				}
+
+				errorType, ok := err.(awserr.RequestFailure)
+
+				if !ok {
+					t.Errorf("Expected awserr.RequestFailure, got %v", errorType)
+				}
+
+			})
+		}
+	})
+
+	t.Run("seek from other starting position", func(t *testing.T) {
+		fixtures := []struct {
+			desc          string
+			initialOffset int
+			offset        int64
+			whence        int
+			expected      int64
+		}{
+			{"whence SeekStart", 3, 2, io.SeekStart, 2},
+			{"whence SeekCurrent", 3, 3, io.SeekCurrent, 6},
+			{"whence SeekEnd", 3, -1, io.SeekEnd, int64(len(content)) - 1},
+		}
+
+		for _, f := range fixtures {
+			f := f
+			t.Run(f.desc, func(t *testing.T) {
+				testFs := s3fs.New(s3cl, *bucket)
+				data, err := testFs.Open(testFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				readBuffer := make([]byte, f.initialOffset)
+				readBytes, err := data.Read(readBuffer)
+				if err != nil {
+					t.Fatal(err)
+				}
+				if readBytes != f.initialOffset {
+					t.Errorf("Read failed during test setup")
+				}
+
+				actual, err := data.(io.Seeker).Seek(f.offset, f.whence)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				if actual != f.expected {
+					t.Errorf("Expected %d, got %d", f.expected, actual)
+				}
+
+			})
+		}
+	})
+
+	t.Run("seek then read", func(t *testing.T) {
+		fixtures := []struct {
+			desc      string
+			readBytes int
+			offset    int64
+			whence    int
+			expected  []byte
+		}{
+			{"whence SeekStart", 3, 2, io.SeekStart, content[2:5]},
+			{"whence SeekStart overflow", 3, 5, io.SeekStart, content[5:7]},
+			{"whence SeekCurrent", 3, 4, io.SeekCurrent, content[4:7]},
+			{"whence SeekEnd", 3, -3, io.SeekEnd, content[len(content)-3:]},
+		}
+
+		for _, f := range fixtures {
+			f := f
+			t.Run(f.desc, func(t *testing.T) {
+				testFs := s3fs.New(s3cl, *bucket)
+				data, err := testFs.Open(testFile)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				_, err = data.(io.Seeker).Seek(f.offset, f.whence)
+				if err != nil {
+					t.Fatal(err)
+				}
+
+				readBuffer := make([]byte, f.readBytes)
+				readBytes, err := data.Read(readBuffer)
+				if readBytes != len(f.expected) {
+					t.Errorf("Read returned unexpected number of bytes")
+				}
+				if err != nil && err != io.EOF {
+					t.Fatal(err)
+				}
+				if bytes.Compare(readBuffer[:readBytes], f.expected) != 0 {
+					t.Errorf("Expected %s, got %s", string(f.expected), string(readBuffer))
+				}
+
+			})
+		}
+	})
+
+}
+
```

Comment:
I think we are missing a tes where we seek to the end and read - we should be getting EOF

Reply:
Added